d2ir/compile.go

```diff
@@ -92,7 +94,133 @@ func (c *compiler) compileClasses(m *Map) {
 			l.Fields = append(l.Fields, base)
 		}
 
-		c.compileClasses(l)
+		c.overlayClasses(l)
+	}
+}
+
+func (c *compiler) compileSubstitutions(m *Map, varsStack []*Map) {
+	for _, f := range m.Fields {
+		if f.Name == "vars" && f.Map() != nil {
+			varsStack = append([]*Map{f.Map()}, varsStack...)
+		}
+		if f.Primary() != nil {
+			c.resolveSubstitutions(varsStack, f.LastRef().AST(), f.Primary())
+		}
+		if f.Map() != nil {
+			// this map could be the "vars"
+			// and if it is, then its already been compiled, so the varsStack can be truncated
+			if f.Name == "vars" {
+				c.compileSubstitutions(f.Map(), varsStack[1:])
+			} else {
+				c.compileSubstitutions(f.Map(), varsStack)
+			}
+		}
+	}
+	for _, e := range m.Edges {
+		if e.Primary() != nil {
+			c.resolveSubstitutions(varsStack, e.LastRef().AST(), e.Primary())
+		}
+		if e.Map() != nil {
+			c.compileSubstitutions(e.Map(), varsStack)
+		}
+	}
+}
+
+func (c *compiler) resolveSubstitutions(varsStack []*Map, node d2ast.Node, scalar *Scalar) {
+	var subbed bool
+	var resolvedField *Field
+
+	switch s := scalar.Value.(type) {
+	case *d2ast.UnquotedString:
+		for i, box := range s.Value {
+			if box.Substitution != nil {
+				for _, vars := range varsStack {
+					resolvedField = c.resolveSubstitution(vars, box.Substitution)
+					if resolvedField != nil {
+						break
+					}
+				}
+				if resolvedField == nil {
+					c.errorf(node, `could not resolve variable "%s"`, strings.Join(box.Substitution.IDA(), "."))
+					return
+				}
+				if resolvedField.Composite != nil {
+					c.errorf(node, `cannot reference map variable "%s"`, strings.Join(box.Substitution.IDA(), "."))
+					return
+				}
+				// If lone and unquoted, replace with value of sub
+				if len(s.Value) == 1 {
```

Comment:
Why bother with this?

Reply:
it's for this test: https://github.com/terrastruct/d2/pull/1473/files#diff-ff6179ec34b122cf73328d5524adaf22e6049d777051407b3f9dac1da530cea5R3309

(if link breaks, search for "quoted-var")

the variable is a quoted string, but the substitution is unquoted string. the contents won't match the type

Reply:
Right but that doesn't matter at the IR stage. I'm fairly certain you could remove this if clause and that test would still pass.

Reply:
the test does not if i remove this.

Reply:
i suspect if you set unquotedstring to a string with quotes, something interpreting the string just gets rid of it, maybe in formatter. i can look deeper if you're not satisfied

Reply:
I'll look into it. I don't see how the formatter can be involved.

Reply:
Here's the right code:

```diff
diff --git a/d2ir/compile.go b/d2ir/compile.go
index f4f0e7d4c..6171b4cc2 100644
--- a/d2ir/compile.go
+++ b/d2ir/compile.go
@@ -195,13 +195,8 @@ func (c *compiler) resolveSubstitutions(varsStack []*Map, node Node) {
 						n.Primary_ = nil
 					}
 				} else {
-					if len(s.Value) == 1 {
-						// If lone and unquoted, replace with value of sub
-						node.Primary().Value = resolvedField.Primary().Value
-					} else {
-						s.Value[i].String = go2.Pointer(resolvedField.Primary().String())
-						subbed = true
-					}
+					s.Value[i].String = go2.Pointer(resolvedField.Primary().Value.ScalarString())
+					subbed = true
 				}
 				if resolvedField.Composite != nil {
 					switch n := node.(type) {
@@ -237,7 +232,7 @@ func (c *compiler) resolveSubstitutions(varsStack []*Map, node Node) {
 					c.errorf(node.LastRef().AST(), `cannot substitute map variable "%s" in quotes`, strings.Join(box.Substitution.IDA(), "."))
 					return
 				}
-				s.Value[i].String = go2.Pointer(resolvedField.Primary().String())
+				s.Value[i].String = go2.Pointer(resolvedField.Primary().Value.ScalarString())
 				subbed = true
 			}
 		}
```

Three tests broke but it looks like they were testing for incorrect behaviour.

Reply:
oh i see. calling `String` is what invoked the formatter.